### PR TITLE
feat: Enable show/hide keyboard shortcut

### PR DESCRIPTION
**Pull request by TheGB0077**
### Second of all, I decided to draft a shortcut to hide and show the app to make it easier to just run it on the background and then use whenever needed, I put it Ctrl + L just to be able to clean the terminal and trigger the event at the same time, but I suppose there's a better alternative! I can work on the details with your feedback.

### CODE REVIEW

**Comment by Merrit** [comment]
Hey there! So glad you've found the app helpful! :bowing_woman:

I will take a look at the PR soon as I get a little time. :grinning:

**Comment by TheGB0077** [comment]
> Hey there! So glad you've found the app helpful! 🙇‍♀️
> 
> I will take a look at the PR soon as I get a little time. 😀

No problem! For a second there I forgot the App was on Android too, but with the latest commit it shouldn't be an issue.

If you want to look into it, the extra dependency added was this:
https://pub.dev/packages/hid_listener

**Comment by Merrit** [comment]
Unfortunately this won't work on Wayland. While looking into this I discovered they did finally add a [global hotkey portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.GlobalShortcuts) however.

Since the app opens quickly and can be bound to any shortcut via the system settings already, do we want to add built-in shortcuts that would exclude a large portion of Linux users?

I am wondering if it would be better to do something like add support for the xdg portal which will be the standard going forward; perhaps by updating the linux support in [hotkey_manager](https://pub.dev/packages/hotkey_manager). 🤔 💭

**Comment by TheGB0077** [comment]
But that solution wouldn't be work on the Windows version of the App, would it? I particulaly took an approach considering the Windows and potential Mac version of the app, if it were to be avaible in the future. Relying on this GlobalShortcuts API wouldn't make the the feature dependent on flatpak? 🤔

As you said, the feature will won't work on wayland, but it looks like it doesn't segfaults, at least on this test I did on a Nobara linux VM:
 
image

Also, this picture kinda teases it, but I packaged a rpm package (and a deb too) for the App. 😊

**Comment by Merrit** [comment]
The xdg portal would only be for the Linux implementation; `hid_listener` or `hotkey_manager` both already work on Windows & Mac; and the xdg portal isn't specific to Flatpak, it is a generic way to do a whole lot of things on Linux (eg file chooser, screenshots, printing, etc). :grin:

**Comment by Merrit** [comment]
I am thinking this might be a good thing to go ahead with for the time being, and the app can do a run-time check to see if it is running under Wayland, and if so disable/hide things related to the built-in hotkey. Can worry about an implementation for Wayland later.

Had a couple of thoughts on it:

- Have you tested if the shortcut works on Windows? I've only tested Linux so far, as it is what I personally use.
- Implementation should be organized outside of `main`, maybe something like a `HotkeyService` class in `lib/src/hotkey` or something like that..
- Default shortcut should be `Meta` + `.` - which seems to be almost a standard hotkey for a system emoji pickers.
- What happens if such a shortcut is already set by the system; Crash, error, etc?
- There should be a section in the settings to allow the user to choose a different hotkey, or to disable the in-app hotkey.
- Should the default hotkey be enabled by default? I am thinking, if the system doesn't already use this hotkey yes, but just wondering if that might be annoying for some. :woman_shrugging: 

What do you think? :smile:

**Comment by TheGB0077** [comment]
Hey @Merrit, thanks again for the Feedback. I've been trying to catch up to reading on the matter of Linux's portals. I'm newish to Linux (only have been using it for 3 months) and still am unfamiliar with several aspects of the internal workings of certain things, but I'm most definitively eager to learn more about it, so if you can point me to some resources about it, I'll appreciate it very much!

Now, regarding your questions/suggestions (In order):

- I haven't tried Windows, as I don't have a Windows with development dependencies.
- Agreed.
- I did some testing for Super + Period and on my Pop_OS system the Super key appears to be captured by default by the App Launcher, making it not work under any scenario as a shortcut. I even tried `RawKeyboard.instance.keysPressed.contains(LogicalKeyboardKey.meta)`, but it simply does not Map into the current pressed keys. (Tested with hid_listener and hotkey_manager). For now, I configured it to Alt + Period.
- I observed two different behaviors, with hotkey_manager the assigned shortcut will override the system shortcut (Linux), while hid_listener just attaches the listener and executes the system shortcut function as well.
- I think so.
- I was thinking about an option to start the App hidden, such that would use this shortcut by default. Something like a toggle in the settings and/or an "--start-hidden" argument.

**Comment by TheGB0077** [comment]
The windows artifact could be useful for testing, can you post a download link for it?

**Comment by Merrit** [comment]
https://snapcraft.io/docs/xdg-desktop-portals
https://github.com/flatpak/xdg-desktop-portal
https://flatpak.github.io/xdg-desktop-portal/

These are the primary places I have researched the xdg portals, including referencing the xml dbus definitions in the github repo.

> I was thinking about an option to start the App hidden, such that would use this shortcut by default. Something like a toggle in the settings and/or an "--start-hidden" argument.

It actually already starts hidden, and programmatically [shows the window once initialization is finished](https://github.com/Merrit/feeling_finder/blob/fe2b06d043d7065ce1b3cd6869833ec8a84e8529/lib/main.dart#L85). So I think this would just be an option in the settings, defaulting to showing on start but allowing the user to toggle "Start hidden" where the app just doesn't run the show command if that preference is set.

[windows-artifacts(2).zip](https://github.com/Merrit/feeling_finder/files/11667489/windows-artifacts.2.zip)

**Comment by TheGB0077** [comment]
So, I meant to ask before doing any changes to the CI. The preferred way to do it would be hotkey_manager, even though it ends up shipping an extra dependency for Linux?  I attached both methods in my most recent commit to open up that for debate.

**Comment by Merrit** [comment]
I'm not familiar enough with either hotkey package to say which is better, that would require research or testing to see if they provide the needed features & work as expected. :)

Regardless, I updated the CI so it won't fail to perform its checks properly at least. :fireworks:

**Comment by TheGB0077** [comment]
I opted to remove the hid_listener. There are quite a few scenarios where it could be quite annoying, so I deleted the code that referenced it.

Now I just had a minor issue due to the new Flutter version making the onClose signal not to be emitted, so the HotKey ends up not being unregistered properly, which ends up causing a segmentation fault whenever the App closes with through the close button (Shortcut event like ctrl + q are covered). Although there is an unmerged fix for that on the Window_Manager package.

**Comment by TheGB0077** [comment]
I used the forked version with the fix for this build, once again I wish to test the Windows artifact to check if everything runs fine :)

File: l10n.yaml [hunk 1]
```diff
@@ -1,3 +1,5 @@
 arb-dir: lib/src/localization
+output-dir: lib/src/localization/gen
 template-arb-file: app_en.arb
 output-localization-file: app_localizations.dart
+synthetic-package: false
```
Comment (Merrit) [inline l10n.yaml:5]: What was the purpose of this change?

File: lib/main.dart [hunk 2]
```diff
@@ -26,6 +28,8 @@ void main(List<String> args) async {
 
   await closeExistingSessions();
 
+  if(platformIsDesktop()) hotKeyService.initHotkeyRegistration();
```
Comment (Merrit) [inline lib/main.dart]: This should happen after the logger is initialized, so the service can use the
logger if needed.

**Comment by Merrit** [comment]
Seems like the debug build is too large to post here. You may wish to build locally. 🫠

File: l10n.yaml (hunk 1 above)
Comment (TheGB0077) [reply l10n.yaml:5]: When I swapped the Window_manger for the forked version the localization files were not being generated anymore, so I did the manual steps. I included flutter_gen, changed it to generate the files inside the output_dir with those two lines. 
I didn't quite like this unreliable behavior so I also created a "suggestion issue" to evaluate Slang as an i18n solution.

File: lib/main.dart (hunk 2 above)
Comment (TheGB0077) [reply lib/main.dart]: Noted!

**Comment by TheGB0077** [comment]
Hey there! I tested the older version of the windows build that you posted here, and it looks like the shortcut it's not working. It might be down the Window Manager library. Can you check if the latest CI build also can use the feature on your Windows machine?

If it does work, I think it's good to merge, otherwise we could perhaps consider this to be Linux only.

**Comment by Merrit** [comment]
I checked and I can confirm it doesn't seem to work on Windows.

I am fine with leaving the feature Linux-only until that gets figured out, if you want to amend it to only initialize / show in settings for Linux. :+1:

**Comment by Merrit** [comment]
LGTM

Thank you for your contribution! :tada: :grin:

**Pull request merged**